artemis-commons/src/main/java/org/apache/activemq/artemis/utils/ByteUtil.java

```diff
@@ -205,23 +205,49 @@ public static int bytesToInt(byte[] b) {
             | ((int) b[0] & 0xff) << 24;
    }
 
+   public static long bytesToLong(byte[] b) {
+      return ((long) b[7] & 0xff)
+         | ((long) b[6] & 0xff) << 8
+         | ((long) b[5] & 0xff) << 16
+         | ((long) b[4] & 0xff) << 24
+         | ((long) b[3] & 0xff) << 32
+         | ((long) b[2] & 0xff) << 40
+         | ((long) b[1] & 0xff) << 48
+         | ((long) b[0] & 0xff) << 56;
+   }
+
    public static byte[] longToBytes(long value) {
       byte[] output = new byte[8];
       longToBytes(value, output, 0);
       return output;
    }
 
    public static void longToBytes(long x, byte[] output, int offset) {
-      output[offset] = (byte)(x >> 56);
-      output[offset + 1] = (byte)(x >> 48);
-      output[offset + 2] = (byte)(x >> 40);
-      output[offset + 3] = (byte)(x >> 32);
-      output[offset + 4] = (byte)(x >> 24);
-      output[offset + 5] = (byte)(x >> 16);
-      output[offset + 6] = (byte)(x >>  8);
+      output[offset] = (byte)(x >>> 56);
+      output[offset + 1] = (byte)(x >>> 48);
+      output[offset + 2] = (byte)(x >>> 40);
+      output[offset + 3] = (byte)(x >>> 32);
+      output[offset + 4] = (byte)(x >>> 24);
+      output[offset + 5] = (byte)(x >>> 16);
+      output[offset + 6] = (byte)(x >>>  8);
```

Comment:
this was a bug originally... >> in Java is signaled.. if you had a negative value things would be pretty much broken.

Reply:
Its true >> and >>> can differ, and that can make a difference depending on their usage, but I don't believe it does in this case though.

This repeatedly carves off the low order 8 bits using a cast to byte, after shifting the original bits by between 56 and 0 steps right, to incrementally capture the original 64 bits from high to low. It discards the actual shifted value at each intermediate step, operating on the original long for the next shift done (if any).

That seems to mean it will neither access or store any of the high order bits introduced by the shifts. Those are what can be influenced by use of >> vs >>>, which would mean this does exactly the same as it did to begin with.

Reply:
(To be clear, >>> usage is fine and if the resulting behaviour hasnt altered then there is obviously no problem from the change being made....I made the comment on the change before I later realised it was already pushed, and was mainly just noting it didnt seem like a necessary change, as smaller changes are nice and easier to review and reason about later)

Reply:
@gemmellr it would been broken without the >>>

>> in Java would add further -1 if the value was negative.. totally different from the intention here.

As the byte was really unsigned, we can't use >> here

As a matter of fact that's how Netty does on their libraries as well.


It wasn't really part of this PR.. it was here as supporting my test. I'm updating the branch now.

Reply:
I still don't see that the way it is used leaves room for a difference, it never seems to store the value or access the bits that the shifting makes a difference to.

Regardless, if you found a defect, add a test for it when you resolve it, makes it clearer what is being fixed and can save discussions like this.